edk2basetools/VfrCompiler/IfrUtility.py

```diff
@@ -0,0 +1,2238 @@
+                    f.write("\t\t\t+{:0>8d}[{:0>8x}] {} <{}>\n".format(FNode.Offset, FNode.BitOffset, FNode.FieldName, FNode.FieldType.TypeName))
+                else:
+                    f.write("\t\t\t+{:0>8d}[{:0>8x}] <{}>\n".format(FNode.Offset, FNode.Offset, FNode.FieldType.TypeName))
+                    f.write("\t\t\t+{:0>8d}[{:0>8x}] <{}>\n".format(FNode.Offset, FNode.BitOffset, FNode.FieldType.TypeName))
+                FNode = FNode.Next
+            f.write("\t\t};\n")
+            f.write("---------------------------------------------------------------\n")
+            pNode = pNode.Next
+        f.write("***************************************************************\n")
+
+
+class SVfrDefaultStoreNode(object):
+    def __init__(self, ObjAddr=None, RefName="", DefaultStoreNameId=0, DefaultId=0):
+        self.ObjAddr = ObjAddr
+        self.RefName = RefName
+        self.DefaultStoreNameId = DefaultStoreNameId
+        self.DefaultId = DefaultId
+        self.Next = None
+
+
+class VfrDefaultStore(object):
+    def __init__(self):
+        self.DefaultStoreList = None
+
+    def Clear(self):
+        self.DefaultStoreList = None
+
+    def RegisterDefaultStore(self, ObjAddr: EFI_IFR_DEFAULTSTORE, RefName, DefaultStoreNameId, DefaultId):
+        if RefName == "" or RefName == None:
+            return VfrReturnCode.VFR_RETURN_FATAL_ERROR
+
+        pNode = self.DefaultStoreList
+        while pNode != None:
+            if pNode.RefName == RefName:
+                return VfrReturnCode.VFR_RETURN_REDEFINED
+            pNode = pNode.Next
+
+        pNode = SVfrDefaultStoreNode(ObjAddr, RefName, DefaultStoreNameId, DefaultId)
+
+        if pNode == None:
+            return VfrReturnCode.VFR_RETURN_OUT_FOR_RESOURCES
+
+        pNode.Next = self.DefaultStoreList
+        self.DefaultStoreList = pNode
+
+        return VfrReturnCode.VFR_RETURN_SUCCESS
+
+    def UpdateDefaultType(self, DefaultNode):
+        pNode = self.DefaultStoreList
+        while pNode != None:
+            if pNode.ObjAddr.DefaultId == DefaultNode.Data.GetDefaultId():
+                DefaultNode.Data.SetType(pNode.RefName)
+            pNode = pNode.Next
+
+    def DefaultIdRegistered(self, DefaultId):
+        pNode = self.DefaultStoreList
+        while pNode != None:
+            if pNode.DefaultId == DefaultId:
+                return True
+            pNode = pNode.Next
+
+        return False
+
+    def ReRegisterDefaultStoreById(self, DefaultId, RefName, DefaultStoreNameId):
+        pNode = self.DefaultStoreList
+        while pNode != None:
+            if pNode.DefaultId == DefaultId:
+                break
+            pNode = pNode.Next
+
+        if pNode == None:
+            return None, VfrReturnCode.VFR_RETURN_UNDEFINED
+        else:
+            if pNode.DefaultStoreNameId == EFI_STRING_ID_INVALID:
+                pNode.DefaultStoreNameId == DefaultStoreNameId
```

Comment:
I think this is wrong. I'm not sure if you are trying to set pNode.DefaultStoreNameId, or do a second comparison, but something is up here.